core/src/main/java/com/linecorp/armeria/server/Http1RequestDecoder.java

```diff
@@ -438,6 +364,124 @@ private void fail(int id, @Nullable RequestHeaders headers, HttpStatus status,
+            req = null;
+            return;
+        }
+        assert req instanceof DecodedHttpRequestWriter;
+        final DecodedHttpRequestWriter decodedReq = (DecodedHttpRequestWriter) req;
+        final DecoderResult decoderResult = content.decoderResult();
+        if (!decoderResult.isSuccess()) {
+            fail(id, decodedReq.headers(), HttpStatus.BAD_REQUEST, "Decoder failure", null);
+            final ProtocolViolationException cause =
+                    new ProtocolViolationException(decoderResult.cause());
+            decodedReq.close(HttpStatusException.of(HttpStatus.BAD_REQUEST, cause));
+            return;
+        }
+
+        final ByteBuf data = content.content();
+        final int dataLength = data.readableBytes();
+        if (dataLength != 0) {
+            decodedReq.increaseTransferredBytes(dataLength);
+            final long maxContentLength = decodedReq.maxRequestLength();
+            final long transferredLength = decodedReq.transferredBytes();
+            if (maxContentLength > 0 && transferredLength > maxContentLength) {
+                final ContentTooLargeException cause =
+                        ContentTooLargeException.builder()
+                                                .maxContentLength(maxContentLength)
+                                                .contentLength(req.headers())
+                                                .transferred(transferredLength)
+                                                .build();
+                discarding = true;
+                req = null;
+                final boolean shouldReset = shouldReset(endOfStream);
+
+                // Wrap the cause with the returned status to let LoggingService correctly log the
+                // status.
+                final HttpStatusException httpStatusException =
+                        HttpStatusException.of(HttpStatus.REQUEST_ENTITY_TOO_LARGE, cause);
+                if (!decodedReq.isFired()) {
+                    assert decodedReq.needsAggregation();
+                    final StreamingDecodedHttpRequest streamingReq = decodedReq.toAbortedStreaming(
+                            inboundTrafficController, httpStatusException, shouldReset);
+                    streamingReq.fireChannelRead(ctx);
+                } else {
+                    decodedReq.setShouldResetOnlyIfRemoteIsOpen(shouldReset);
+                    decodedReq.abortResponse(httpStatusException, true);
+                }
+                return;
+            }
+
+            if (decodedReq.isOpen()) {
+                decodedReq.write(HttpData.wrap(data.retain()));
+            }
+        }
+
+        if (endOfStream) {
+            final HttpHeaders trailingHeaders = ((LastHttpContent) content).trailingHeaders();
+            if (!trailingHeaders.isEmpty()) {
+                decodedReq.write(ArmeriaHttpUtil.toArmeria(trailingHeaders));
+            }
+            decodedReq.close();
+            if (!decodedReq.isFired()) {
```

Comment:
Optional) I think this class is probably less confusing if we drop the `isFired` check entirely and rely on `req#needsAggregation`. I think this confusion stems from the fact that:

- We already have multiple states to check whether `fireChannelRead` was called (`decodedReq#fired`, `req#needsAggregation`)
- `req.toAbortedStreaming` doesn't check `isFired` anyways when returning 413. This can be confusing when there are multiple `req#toAbortedStreaming` calls like here (`413`, scheduler)

Reply:
> if we drop the isFired check entirely and rely on req#needsAggregation

That's a good suggestion. 👍 But this PR will be closed. 😅